### PR TITLE
feat: add `lapack/base/dlassq`

### PR DESCRIPTION
**Pull request by Pranavchiku**
## Description

> What is the purpose of this pull request?

This pull request add JS implementation for `lapack/base/dlassq`.

## Related Issues

> Does this pull request have any related issues?

Towards https://github.com/stdlib-js/stdlib/issues/2464

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

TODO:

- [x] Add more tests

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/lapack/base/dlassq/README.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+</section>
+
+<!-- /.usage -->
+
+<section class="notes">
+
+## Notes
+
+-   `dlassq()` corresponds to the [LAPACK][LAPACK] level 1 function [`dlassq`][dlassq].
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dlassq/README.md]:
```suggestion
-   `dlassq()` corresponds to the [LAPACK][LAPACK] function [`dlassq`][dlassq].
```

File: lib/node_modules/@stdlib/lapack/base/dlassq/benchmark/benchmark.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2020 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dlassq/benchmark/benchmark.js]:
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/lapack/base/dlassq/benchmark/benchmark.ndarray.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2020 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dlassq/benchmark/benchmark.ndarray.js]:
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+var SBIG = 1.11137937474253874e-162;
+var SSML = 4.49891379454319638e+161;
+var TBIG = 1.99791907220223503e+146;
+var TSML = 1.49166814624004135e-154;
+
+
+// MAIN //
+
+/**
+* Returns the values `scale_out` and `sumsq_out` such that `(scale_out**2)*sumsq_out = x( 1 )**2 +...+ x( n )**2 + (scale**2)*sumsq` where `x( i ) = X( 1 + ( i - 1 )*INCX )`.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js]:
```suggestion
* Returns the values `scale_out` and `sumsq_out` such that `(scale_out**2)*sumsq_out = x( 1 )**2 +...+ x( n )**2 + (scale**2)*sumsq`.
```

I think we can remove this from here and elsewhere except `lib/dlassq.js`, not sure though

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+* @example
+* var Float64Array = require( '@stdlib/array/float64' );
+*
+* var out = new Float64Array( [ 0.0, 0.0 ] );
+* var X = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
+*
+* dlassq( 4, X, 1, 0, 1, 0, out, 1, 0 );
+* // out => <Float64Array>[ 1.0, 30.0 ]
+*/
+function dlassq( N, X, strideX, offsetX, scale, sumsq, out, strideOut, offsetOut ) { // eslint-disable-line max-len
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js:63]:
```suggestion
function dlassq( N, X, strideX, offsetX, out, strideOut, offsetOut ) { // eslint-disable-line max-len
```
Instead of explicitly providing `scale` and `sumsq`, do you wish to pass it via the same array as input and update it accordingly?

cc @kgryte

File: lib/node_modules/@stdlib/lapack/base/dlassq/README.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# dlassq
+
+> Return the values `scale_out` and `sumsq_out` such that `(scale_out**2)*sumsq_out = x( 1 )**2 +...+ x( n )**2 + (scale**2)*sumsq` where `x( i ) = X( 1 + ( i - 1 )*INCX )`
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/README.md]: This description needs to be in plain English.

File: lib/node_modules/@stdlib/lapack/base/dlassq/README.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var dlassq = require( '@stdlib/lapack/base/dlassq' );
+var Float64Array = require( '@stdlib/array/float64' );
+
+var out = new Float64Array( [ 0.0, 0.0 ] );
+var X = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
+
+dlassq( 4, X, 1, 1, 0, out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/README.md]:
```suggestion
dlassq( 4, X, 1, 1.0, 0.0, out );
```

File: lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        Input array.
+
+    incX: integer
+        Stride length for `X`.
+
+    scale: number
+        Scaling factor for the sum of squares.
+
+    sumsq: number
+        Basic sum of squares from which output is
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt]: Text wrapping is off.

File: lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt [hunk 9]
```diff
@@ -0,0 +1,97 @@
+
+{{alias}}( N, X, incX, scale, sumsq, out )
+    Returns the values `scale_out` and `sumsq_out` such that
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt]: Plain English.

File: lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    Examples
+    --------
+    > var out = new {{alias:@stdlib/array/float64}}( [ 0.0, 0.0 ] );
+    > var X = new {{alias:@stdlib/array/float64}}( [ 1.0, 2.0, 3.0, 4.0 ] );
+    > {{alias}}( 4, X, 1, 1, 0, out )
+    <Float64Array>[ 1.0, 30.0 ]
+
+
+{{alias}}.ndarray( N, X, sx, ox, scale, sumsq, out, sout, oout )
+    Returns the values `scale_out` and `sumsq_out` using alternative
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt]: Plain English.

File: lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        Stride length for `X`.
+
+    ox: integer
+        Starting index for `X`.
+
+    scale: number
+        Scaling factor for the sum of squares.
+
+    sumsq: number
+        Basic sum of squares from which output is
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/docs/repl.txt]: Text wrapping.

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js (hunk 4 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js]: Yes, that doesn't make sense, as `INCX` is not applicable and also doesn't match our naming conventions.

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js (hunk 5 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js:63]: No, they should be passed separately as scalar values. The output array should be considered distinct. If someone wants to overwrite, they can do

```javascript
dlassq( N, X, strideX, offsetX, out[0], out[1], out, strideOut, offsetOut )
```

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+		}
+	}
+	// Combine abig and amed or amed and asml if more than one accumulator was used.
+	if ( abig > 0.0 ) {
+		// Combine abig and amed if abig > 0.
+		if ( amed > 0.0 ) {
+			abig += ( amed * SBIG ) * SBIG;
+		}
+		scale = 1.0 / SBIG;
+		sumsq = abig;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js]: I believe you need to set output values in this branch.

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+			} else {
+				ymin = asml;
+				ymax = amed;
+			}
+			scale = 1.0;
+			sumsq = pow( ymax, 2 ) * ( 1.0 + pow( ymin / ymax, 2 ) );
+		} else {
+			scale = 1.0 / SSML;
+			sumsq = asml;
+		}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/base.js:172]: I believe you need to set output values in this branch.

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/dlassq.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+*
+* var out = new Float64Array( [ 0.0, 0.0 ] );
+* var X = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
+*
+* var out = dlassq( 4, X, 1, 1, 0, out );
+* // out => <Float64Array>[ 1.0, 30.0 ]
+*/
+function dlassq( N, X, incX, scale, sumsq, out ) {
+	var ox;
+	ox = 0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/dlassq.js]:
```suggestion
	var ox = 0;
```

In situations such as this, you can combine declaration and assignment.

File: lib/node_modules/@stdlib/lapack/base/dlassq/lib/dlassq.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+
+// MAIN //
+
+/**
+* Returns the values `scale_out` and `sumsq_out` such that `(scale_out**2)*sumsq_out = x( 1 )**2 +...+ x( n )**2 + (scale**2)*sumsq` where `x( i ) = X( 1 + ( i - 1 )*INCX )`.
+*
+* @param {NonNegativeInteger} N - number of elements to be used from `X`
+* @param {Float64Array} X - input array
+* @param {integer} incX - stride length for `X`
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dlassq/lib/dlassq.js]: Typically, we use `strideX`, rather than `incX`, for consistency elsewhere.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks for working on this, @Pranavchiku.

I ended up having to do a decent amount of clean-up. We'll want to work on reducing that moving forward.

**Comment by Pranavchiku** [comment]
I am sorry for making you put in lots of efforts in PR reviews. I'll take a close look the changes you made and assure quality PRs from now.
